tests/integration_tests/test_botorch.py

```diff
@@ -84,6 +84,60 @@ def candidates_func(
     assert candidates_func_call_count == n_trials - n_startup_trials
 
 
+@pytest.mark.parametrize("candidates_func_kind", ["qei", "qehvi", "qparego", "qnhevi"])
+@pytest.mark.parametrize("is_constrained", [True, False])
+def test_botorch_specify_candidates_func(candidates_func_kind: str, is_constrained: bool) -> None:
+
+    n_trials = 4
+    n_startup_trials = 2
+
+    if candidates_func_kind == "qei":
+        candidates_func = integration.botorch.qei_candidates_func
+        n_objectives = 1
+    elif candidates_func_kind == "qehvi":
+        candidates_func = integration.botorch.qehvi_candidates_func
+        n_objectives = 2
+    elif candidates_func_kind == "qparego":
+        candidates_func = integration.botorch.qparego_candidates_func
+        n_objectives = 4
+    elif candidates_func_kind == "qnhevi":
+        candidates_func = integration.botorch.qnehvi_candidates_func
+        n_objectives = 2
+    else:
+        assert False
+
+    if is_constrained:
+        constraints_func_call_count = 0
+
+        def constraints_func(trial: FrozenTrial) -> Sequence[float]:
+            xs = sum(trial.params[f"x{i}"] for i in range(n_objectives))
+
+            nonlocal constraints_func_call_count
+            constraints_func_call_count += 1
+
+            return (xs - 0.5,)
+
+        sampler = BoTorchSampler(
+            constraints_func=constraints_func,
+            candidates_func=candidates_func,
+            n_startup_trials=n_startup_trials,
+        )
+    else:
+        sampler = BoTorchSampler(
+            candidates_func=candidates_func,
+            n_startup_trials=n_startup_trials,
+        )
```

Comment:
If the process differs greatly due to such different arguments (`is_constrained = True` or `False`), it is rather better to split the test.

Reply:
Agreed. I split the test function.